internal/core/operations/healthcontroller/executor.go

```diff
@@ -369,3 +385,122 @@ func (ex *Executor) canPerformDownscale(ctx context.Context, scheduler *entities
+	if len(roomsWithPreviousSchedulerVersion) < maxRoomsToSurge {
+		maxRoomsToSurge = len(roomsWithPreviousSchedulerVersion)
+	}
+	if maxRoomsToSurge <= 0 {
+		maxRoomsToSurge = 1
+	}
+	logger.Info(
+		"upscaling new rooms",
+		zap.Int("desired", desiredNumberOfRooms),
+		zap.Int("maxSurgeAmount", maxSurgeAmount),
+		zap.Int("available", len(availableRoomsIDs)),
+		zap.Int("oldRooms", len(roomsWithPreviousSchedulerVersion)),
+		zap.Int("maxRoomsToSurge", maxRoomsToSurge),
+	)
+	addOp, err := ex.operationManager.CreatePriorityOperation(ctx, op.SchedulerName, &add.Definition{
+		Amount: int32(maxRoomsToSurge),
+	})
+	if err != nil {
+		logger.Error("failed to enqueue add operation for rolling update", zap.Error(err))
+		return err
+	}
+	msgToAppend := fmt.Sprintf("created operation (id: %s) to surge %v rooms.", addOp.ID, maxSurgeAmount)
+	ex.operationManager.AppendOperationEventToExecutionHistory(ctx, op, msgToAppend)
+	ex.setTookAction(def, true)
+
+	roomsMarkedForDeletion, err := ex.markPreviousSchedulerRoomsForDeletion(
+		ctx,
+		logger,
+		scheduler,
+		roomsWithPreviousSchedulerVersion,
+		desiredNumberOfRooms,
+	)
+	if err != nil {
+		logger.Error("could not delete rooms with previous scheduler version", zap.Error(err))
+		return err
+	}
+	if len(roomsMarkedForDeletion) <= 0 {
+		return nil
+	}
+	removeOp, err := ex.operationManager.CreateOperation(ctx, op.SchedulerName, &remove.Definition{
+		RoomsIDs: roomsMarkedForDeletion,
+	})
+	if err != nil {
+		logger.Error("failed to enqueue remove operation for rolling update", zap.Error(err))
+		return err
+	}
+	msgToAppend = fmt.Sprintf("created operation (id: %s) to remove rooms with previous scheduler version.", removeOp.ID)
+	ex.operationManager.AppendOperationEventToExecutionHistory(ctx, op, msgToAppend)
+	ex.setTookAction(def, true)
+
+	return nil
+}
+
+func (ex *Executor) markPreviousSchedulerRoomsForDeletion(
+	ctx context.Context,
+	logger *zap.Logger,
+	scheduler *entities.Scheduler,
+	roomsWithPreviousSchedulerVersion []string,
+	desiredNumberOfRooms int,
+) ([]string, error) {
+	logger.Debug(
+		"ensuring only rooms with current scheduler version are running",
+		zap.String("scheduler.Name", scheduler.Name),
+		zap.String("scheduler.Version", scheduler.Spec.Version),
+	)
+	readyRoomIDs, err := ex.roomStorage.GetRoomIDsByStatus(ctx, scheduler.Name, game_room.GameStatusReady)
+	if err != nil {
+		return []string{}, fmt.Errorf("failed to list scheduler rooms on ready status: %w", err)
+	}
+	bufferRoomsToBeRemoved := len(readyRoomIDs) - desiredNumberOfRooms
+	if bufferRoomsToBeRemoved < 0 {
+		logger.Sugar().Debugf("can not delete old rooms without offending autoscale: %d", bufferRoomsToBeRemoved)
+		return []string{}, nil
+	}
+	if bufferRoomsToBeRemoved > len(roomsWithPreviousSchedulerVersion) {
+		return roomsWithPreviousSchedulerVersion, nil
+	}
+	return roomsWithPreviousSchedulerVersion[:bufferRoomsToBeRemoved], nil
```

Comment:
should we consider the down threshold here?

Reply:
It's an overkill. With this computation, I'm only deleting rooms considering the `ready` amount, it's similar to Kubernetes when you define `maxUnavailable: 0`. We will never go below the desired number of rooms and we always downscale to the desired.

I think that during rolling update, the games don't want to fine-grain control how aggressive we are downscaling the pods if we do it safely. The only downside of this approach is that the process will take longer because we never go below the desired, that's when people generally use `maxUnavailble: 25%` or something like it

Reply:
makes sense :) 

unrelated: sice we do `len(readyRoomIDs) - desiredNumberOfRooms`, can we ever get stuck or take a long time to renew rooms?

Reply:
before scaling and rolling updates were two district actions, right? now, they can be done together (as i understand), if so, it's nice to respect the same behavior if possible

one pro for not doing is the possibility where we never/take too long to renew pods because of down threshold